lib/Backend/arm64/EncoderMD.cpp

```diff
@@ -186,555 +186,497 @@ EncoderMD::DecodeMemoryOpnd(IR::Opnd* opnd, ARM64_REGISTER &baseRegResult, ARM64
-        Linker.SetTarget(Emitter);
-        bytes = EmitBranch(Emitter, Linker, COND_PL);
+        bytes = this->EmitConditionalBranch(Emitter, instr, COND_PL);
         break;
 
     case Js::OpCode::BVS:
-        EncodeReloc::New(&m_relocList, RelocTypeBranch19, m_pc, instr->AsBranchInstr()->GetTarget(), m_encoder->m_tempAlloc);
-        Linker.SetTarget(Emitter);
-        bytes = EmitBranch(Emitter, Linker, COND_VS);
+        bytes = this->EmitConditionalBranch(Emitter, instr, COND_VS);
         break;
     
     case Js::OpCode::BVC:
-        EncodeReloc::New(&m_relocList, RelocTypeBranch19, m_pc, instr->AsBranchInstr()->GetTarget(), m_encoder->m_tempAlloc);
-        Linker.SetTarget(Emitter);
-        bytes = EmitBranch(Emitter, Linker, COND_VC);
+        bytes = this->EmitConditionalBranch(Emitter, instr, COND_VC);
         break;
     
     case Js::OpCode::DEBUGBREAK:
         bytes = EmitDebugBreak(Emitter);
         break;
 
     case Js::OpCode::CLZ:
-        dst = instr->GetDst();
-        Assert(dst->IsRegOpnd());
-        src1 = instr->GetSrc1();
-        Assert(src1->IsRegOpnd());
-        bytes = EmitClz(Emitter, this->GetRegEncode(dst->AsRegOpnd()), this->GetRegEncode(src1->AsRegOpnd()));
+        bytes = this->EmitOp2Register(Emitter, instr, EmitClz, EmitClz64);
         break;
 
-    // ARM64_WORKITEM: Should legalizer change this to SUBS before the encoder?
+    // Legalizer should convert this to SUBS before getting here
     case Js::OpCode::CMP:
-        dst = instr->GetDst();
-        Assert(dst->IsRegOpnd());
-        src1 = instr->GetSrc1();
-        if (src1->IsImmediateOpnd())
-        {
-            immediate = src1->GetImmediateValue(instr->m_func);
-            bytes = EmitSubsImmediate(Emitter, ARMREG_ZR, this->GetRegEncode(dst->AsRegOpnd()), ULONG(immediate));
-        }
-        else
-        {
-            Assert(src1->IsRegOpnd());
-            bytes = EmitSubsRegister(Emitter, ARMREG_ZR, this->GetRegEncode(dst->AsRegOpnd()), this->GetRegEncode(src1->AsRegOpnd()));
-        }
+        Assert(false);
         break;
 
-    // ARM64_WORKITEM: Should legalizer change this to ADDS before the encoder?
+    // Legalizer should convert this to ADDS before getting here
     case Js::OpCode::CMN:
-        dst = instr->GetDst();
-        Assert(dst->IsRegOpnd());
-        src1 = instr->GetSrc1();
-        if (src1->IsImmediateOpnd())
-        {
-            immediate = src1->GetImmediateValue(instr->m_func);
-            bytes = EmitAddsImmediate(Emitter, ARMREG_ZR, this->GetRegEncode(dst->AsRegOpnd()), ULONG(immediate));
-        }
-        else
-        {
-            Assert(src1->IsRegOpnd());
-            bytes = EmitAddsRegister(Emitter, ARMREG_ZR, this->GetRegEncode(dst->AsRegOpnd()), this->GetRegEncode(src1->AsRegOpnd()));
-        }
+        Assert(false);
         break;
 
     case Js::OpCode::CMP_ASR31:
         dst = instr->GetDst();
```

Comment:
CMP compares src1 and src2 and not dst. @sigatrev was trying to fix the same in #3886

Reply:
Ah, sorry for the misunderstanding. Will address this.